### PR TITLE
Get bookings for date API

### PR DESCRIPTION
**Pull request by user84918342**

### CODE REVIEW

File: Book-A-Desk.Api/BookingsHttpHandler.fs [hunk 1]
```diff
@@ -23,12 +23,13 @@ open Book_A_Desk.Infrastructure.DynamoDbEventStore
...
     {
         HandlePostWith: Models.Booking -> HttpHandler
-        HandleGetByEmailAndDate: unit -> HttpHandler
+        HandleGet: unit -> HttpHandler
     }
 
 module BookingsHttpHandler =
     let private notifyBooking output (booking : Models.Booking) notifySuccess errorHandler next context =
         task {
+            // remember me
```
Comment (AlexGagne) [inline Book-A-Desk.Api/BookingsHttpHandler.fs]: I don't think this comment is necessary, you can remove it

Comment (user84918342) [reply Book-A-Desk.Api/BookingsHttpHandler.fs]: Oh, it was for myself so that I could find the place where to comment the lines out that throw the error because I don't have an email configuration. 
Thank you!

Comment (AlexGagne) [reply Book-A-Desk.Api/BookingsHttpHandler.fs]: All good. A good practice is to re-read your code when creating a PR, then you can spot these little comments that you have added during development and remove them before merging :)

File: Book-A-Desk.Domain/QueriesHandler.Reservation.fs [hunk 2]
```diff
@@ -34,6 +35,19 @@ module rec ReservationsQueriesHandler =
         let bookings = (ReservationAggregate.getCurrentStateFrom bookingEvents).BookedDesks
         return List.where (fun booking -> booking.Date >= date) bookings
     }
+
+    let getFilteredBookings (bookingEvents : seq<DomainEvent>) (date : DateTime option)  (officeId : OfficeId option) (email : EmailAddress option) : Result<Booking list, string> = result {
```
Comment (AlexGagne) [inline Book-A-Desk.Domain/QueriesHandler.Reservation.fs:30]: I think we are missing some unit tests to validate this logic. Tests like:

`Given booking events and a date When filtering bookings Then all bookings for the date are returned`

Ideally, we should have 1 test per filter parameter (date, office and email) + tests that have combinations of the filter parameters

File: Book-A-Desk.Infrastructure/DynamoDbEventStore.fs [hunk 3]
```diff
@@ -19,7 +19,7 @@ module rec DynamoDbEventStore =
                 .Create<ReservationEvent>(
                     dynamoDbClient,
                     tableName = "ReservationEvents",
-                    createIfNotExists = false)
```
Comment (AlexGagne) [inline Book-A-Desk.Infrastructure/DynamoDbEventStore.fs]: I think you can undo this change. We did this to make it run locally for you, but it's not something that we want in the deployed code

File: Book-A-Desk.Api/BookingsHttpHandler.fs [hunk 4]
```diff
@@ -99,68 +99,53 @@ module BookingsHttpHandler =
...
-                                    Booking.value booking.OfficeId booking.Date booking.EmailAddress
-                                    )
-                                |> List.toArray
-                                |> fun l -> { Bookings.Items = l }
-                            return! json bookings next context
-                        | Error e ->
-                            context.SetStatusCode(500)
-                            return! text ("Internal Error: " + e) next context
+                    let officeId = InputParser.parseOfficeIdFromContext context
+
```
Comment (AlexGagne) [inline Book-A-Desk.Api/BookingsHttpHandler.fs:113]: Tiny detail: please remove one of the empty spaces in-between `let officeId` and `let eventStore` to keep the code tidy

File: Book-A-Desk.Api/InputParser.fs [hunk 5]
```diff
@@ -25,4 +25,12 @@ module InputParser =
         let wasParsed, officeIdGuid = Guid.TryParse reference
         let officeId = officeIdGuid |> OfficeId        
         if wasParsed then Ok officeId
-        else Error ()
\ No newline at end of file
+        else Error ()
+
+    let parseOfficeIdFromContext (context : HttpContext) =
+        context.TryGetQueryStringValue "office"
+        |> Option.bind (fun (office: string) ->
```
Comment (AlexGagne) [inline Book-A-Desk.Api/InputParser.fs:35]: I think this code is a little weird since we are mixing both options and errors when trying to reuse the existing parseOfficeId. I think a refactor will help here:
1. Change parseOfficeId to return an Option instead of an error.
2. In parseOfficeIdFromContext, you can call parseOfficeId directly
3. In the code that currently calls parseOfficeId, you can handle the Option instead of a Result. You probably will have to map your Option<OfficeId> to a Result<OfficeId, String> in the existing callers.

File: Book-A-Desk.Api/BookingsHttpHandler.fs [hunk 6]
```diff
@@ -99,68 +99,53 @@ module BookingsHttpHandler =
...
+        let handleGet eventStore date officeId email = asyncResult {
             let! bookingEvents = eventStore.GetEvents ()
-            return! ReservationsQueriesHandler.getUserBookingsStartFrom bookingEvents email date
-        }
-        
-        let handleGetByDateFromEventStore eventStore date = asyncResult {
-            let! bookingEvents = eventStore.GetEvents ()
-            return! ReservationsQueriesHandler.getUsersBookingsStartFrom bookingEvents date
+
+            match officeId with
```
Comment (AlexGagne) [inline Book-A-Desk.Api/BookingsHttpHandler.fs]: So, I see why you have changed this code to sometimes call `getFilteredBookings` and sometimes call `getUserBookingsStartFrom` (failing api tests). These tests were implemented when we partially implemented the "Upcoming Bookings" feature. For now, let's ignore this "Upcoming Bookings" feature, we will revisit it later.

For now, I think you can do this:
1. We will reimplement "Upcoming Bookings" later, so all logic relating to it can be removed. In this case, you can change this `handleGet` to only call `ReservationsQueriesHandler.getFilteredBookings`
2. The API tests will break, however what we will do is to change them to:
    - Remove all tests that talk about "returning bookings of future date".
    - Add new tests to test our filtering (keep them at a high level. Our unit tests are responsible for testing all of the combinations, these tests should test only 1-2 combinations)

When we actually reimplement "Upcoming Bookings", we will add another query parameter (let's call it startFrom for now) to specifically have bookings starting from a certain date

Comment (AlexGagne) [reply Book-A-Desk.Api/BookingsHttpHandler.fs]: Another note:
If we ever run into the same situation, where some parameters mean we should call different logic, this check should be done in the Domain. The Domain should handle all logic, the Api Handlers should only do two things:
1. Call the Infrastructure code (to get the Booking Events)
2. Call the Domain
3. Transform types to and from the API definition and handle API-related concepts like error codes.

File: Book-A-Desk.Infrastructure/DynamoDbEventStore.fs (hunk 3 above)
Comment (user84918342) [reply Book-A-Desk.Infrastructure/DynamoDbEventStore.fs]: But doesn't that mean that everyone who tries to run it locally (or just running it for the first time in general) has to create the database themself?

Comment (AlexGagne) [reply Book-A-Desk.Infrastructure/DynamoDbEventStore.fs]: Yes, but it's supposed to be done already in the setup part of the README. The important point is that we do not want the production code to create tables accidentally.

File: Book-A-Desk.Api.Tests/FilteredBookingsTest.fs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+open Book_A_Desk.Infrastructure.DynamoDbEventStore
+open Book_A_Desk.Domain
+open Book_A_Desk.Infrastructure
+
+open Book_A_Desk.Domain.Events
+open Book_A_Desk.Domain.Reservation
+open Book_A_Desk.Domain.Reservation.Queries
+
+open Book_A_Desk.Domain.Reservation.Events
+
```
Comment (AlexGagne) [inline Book-A-Desk.Api.Tests/FilteredBookingsTest.fs]: A general tidyness comment.

Having a tidy codebase is a pretty important thing. It helps readability and understanding but also prevents the Broken Window Theory. Essentially, if a codebase looks unkempt, it's very easy to let the quality of the code go since if nobody put effort into maintaining it, why should you?

I don't mean that your code isn't clean (it's generally pretty good), but there are some patterns that I see that should be fixed. These can sound like nitpicks or very minor details, but I think they do make a difference in the long-term health of the codebase.

Here's a few things that I see:
1. You generally want to keep a maximum of 1 empty line in-between lines of code (ex: lines 18 and 19, lines 35 and 36 violate this rule).
2. If you are specifying the type of a variable or function, please keep one empty space behind and after `:`. Some examples: line 23, line 29, line 52, line 60. So `let createDeskBookedEvent date officeId email : DomainEvent`. This is a F# standard, so it's okay if you're not familiar since you are learning the language.
3. Empty space at the beginning of a function (line 70)
4. Inconsistent spacing in-between variables declared at the top-level of a module (compare line 28-29 with lines 41-43). Choose one and stick with it. For me, I usually keep top-level declarations with 1 space in-beween. For code inside a function, it depends, but I generally group variables together in groups that semantically belong together with 1 space in-between each group.
5. After calling a function, you generally have a space in-between the function name and arguments (example: lines 65 and 66)
6. Inconsistent indentation: See the indentation of `mockProvideEventStore` compared to `createDeskBookedEvent`. The definition of `mockProvideEventStore` is indented by too much

File: Book-A-Desk.Api.Tests/FilteredBookingsTest.fs [hunk 8]
```diff
@@ -0,0 +1,90 @@
+module Book_A_Desk.Api.Tests.FilteredBookingsTest
```
Comment (AlexGagne) [inline Book-A-Desk.Api.Tests/FilteredBookingsTest.fs]: In this project, there are two types of tests:
1. API Tests, which test calling the API directly
2. Unit Tests, which call code directly in an isolated fashion

Currently, the test in this file is kind of a mix of both an API Test and a Unit Test. You are not fully isolating the code (where you would directly call ReservationsQueriesHandler.getFilteredBookings) and you are not calling the API. Let me show you what I mean.

This file is in the API Tests project, so let me show you how to create an API Test. First, we want to setup a TestServer. This TestServer will run an ASP.Net server in-memory and allows you to call the API directly. This allows you to test the integration of the different parts of the system without requiring you to fully run a server. You can see in the file GetBookingsTests.fs in the lines starting with 

``` fsharp
let mockApiDependencyFactory = //...
use httpClient = TestServer.createAndRun mockApiDependencyFactory 
```

What this code does is sets up all of the dependencies of the system (such as the mock provide event store) and directly tests the code all the way from Program.fs to the HttpHandler. The goal is to test the integration of the different parts of the system. Since this might change more often than the rest of the system, we generally write a limited number of API tests, where the goal is to test that all of the code works well together, more than individual parts.

When writing UnitTests, you want to include the files the Domain.Tests project. As well, we generally only unit test the files in the Domain, since this is where the core and complex part of the system should live. When writing unit tests, you provide only what is required to run the test. For ReservationsQueriesHandler.getFilteredBookings, this means a sequence of BookingEvent, a Date Option, OfficeId Option and Email option. How this information is provided (in our code, by the EventStore) is not important. This allows you to test the code very thoroughly since you are not relying on dependencies that could potentially change, making these tests very stable

File: Book-A-Desk.Api.Tests/FilteredBookingsTest.fs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                    Booking.value booking.OfficeId booking.Date booking.EmailAddress)
+                |> List.toArray
+                |> fun l -> { Bookings.Items = l }
+            bookings
+        | Error e -> failwith"Error occured"
+    | None -> failwith"Empty booking events"
+
+[<Fact>]
+let ``GIVEN booking events and a date WHEN filtering bookings THEN all bookings for the date are returned`` () = async {
+
```
Comment (AlexGagne) [inline Book-A-Desk.Api.Tests/FilteredBookingsTest.fs]: You generally want to isolate what's called your "system under test". In this case, this should be calling the API. It should be clear inside your test what is being tested. One way to achieve this is to have clear steps: a setup step, a testing step and an assert step. We generally call these "Arrange, Act and Assert". Unless you have a very long "Arrange" step, it's generally good practice to write your setup code, followed by an empty line, then the test and then the Asserts. 

Something like:
``` fsharp
[<Fact>]
let test () =
    let testSetup1 = 123
    let testSetup2 = 456

    let result = callingTheApi testSetup1 testSetup2

    Assert.Equal(expectedResult, result)
```

File: Book-A-Desk.Api.Tests/FilteredBookingsTest.fs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+let getBookingEvents : Option<seq<DomainEvent>> =
+    async {
+        let! result = mockProvideEventStore.GetEvents()
+        match result with
+        | Ok events -> return Some events
+        | Error errorMessage -> return None
+    }
+    |> Async.RunSynchronously
+
+let getBookingsFrom (bookingEvents: Option<seq<DomainEvent>>) (date : DateTime option)  (officeId : OfficeId option) (email : EmailAddress option) =
```
Comment (AlexGagne) [inline Book-A-Desk.Api.Tests/FilteredBookingsTest.fs]: I'm not quite sure I understand what this code is doing. It seems to be a copy-paste from the BookingsHttpHandler. I suppose by doing this you are testing the code in the BookingsHttpHandler, but it means you would have to modify the test every time you modify the BookingsHttpHandler. This is quite bad because it is easy for someone to forget and all of a sudden, you are no longer running valid tests.

To fix this, you need to call the TestServer with the API call directly. You can see in GetBookingsTests.fs how to do it, but don't hesitate to ask for clarification if needed.

File: Book-A-Desk.Api.Tests/FilteredBookingsTest.fs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    ]
+        
+
+let mockProvideEventStore =
+        {
+            GetEvents = fun _ -> bookings |> Ok |> async.Return
+            AppendEvents = fun _ -> failwith "should not be called"
+        } : DynamoDbEventStore.DynamoDbEventStore
+
+let getBookingEvents : Option<seq<DomainEvent>> =
```
Comment (AlexGagne) [inline Book-A-Desk.Api.Tests/FilteredBookingsTest.fs]: This might be a misunderstanding of F#. This is not a function, it's a variable. The code here is actually run when the API.Tests project is loaded and the code is executed to create a variable called `getBookingEvents`. If you want to turn it into a function, you need to add a `()`. So: `let getBookingEvents () = // the code to run`.

Although in this case, I don't think this code is useful. If you are writing an API Test, the event store should be injected in the `ApiDependencyFactory`, which means you don't need to call it. If you are writing a unit test, you should provide the list of bookings directly in the call to `ReservationsQueriesHandler.getFilteredBookings`, without going through a dependency or function.

File: Book-A-Desk.Api.Tests/FilteredBookingsTest.fs (hunk 7 above)
Comment (AlexGagne) [reply Book-A-Desk.Api.Tests/FilteredBookingsTest.fs]: To reiterate: your code is generally pretty tidy! I wrote this long text because I want to explain why I added these comments when they can look like meaningless nitpicks. I'd rather overexplain something you might know already then underexplain and have you misunderstand my message :).
